lib/upipe-modules/upipe_row_split.c

```diff
@@ -0,0 +1,357 @@
+            return upipe_row_split_set_flow_def(upipe, flow_def);
+        }
+        case UPIPE_GET_FLOW_DEF:
+        case UPIPE_GET_OUTPUT:
+        case UPIPE_SET_OUTPUT:
+            return upipe_row_split_control_output(upipe, command, args);
+        case UPIPE_SET_OPTION: {
+             const char *k = va_arg(args, const char *);
+             const char *v = va_arg(args, const char *);
+            return upipe_row_split_set_option(upipe, k, v);
+        }
+        default:
+            return UBASE_ERR_UNHANDLED;
+    }
+}
+
+static void upipe_row_split_free(struct upipe *upipe)
+{
+    upipe_row_split_clean_ubuf_mgr(upipe);
+    upipe_row_split_clean_urefcount(upipe);
+    upipe_row_split_clean_output(upipe);
+    upipe_row_split_clean_input(upipe);
+    upipe_row_split_free_void(upipe);
+}
+
+static struct upipe *upipe_row_split_alloc(struct upipe_mgr *mgr,
+                                           struct uprobe *uprobe,
+                                           uint32_t signature,
+                                           va_list args)
+{
+    struct upipe *upipe =
+        upipe_row_split_alloc_void(mgr, uprobe, signature, args);
+    if (unlikely(upipe == NULL))
+        return NULL;
+    struct upipe_row_split *upipe_row_split = upipe_row_split_from_upipe(upipe);
+
+    upipe_row_split->slice_height = 16;
+
+    upipe_row_split_init_urefcount(upipe);
+    upipe_row_split_init_input(upipe);
+    upipe_row_split_init_ubuf_mgr(upipe);
+    upipe_row_split_init_output(upipe);
+
+    return upipe;
+}
+
+static void upipe_row_split_input(struct upipe *upipe, struct uref *uref,
+                                  struct upump **upump_p)
+{
+    if (!upipe_row_split_check_input(upipe)) {
+        upipe_row_split_hold_input(upipe, uref);
+        upipe_row_split_block_input(upipe, upump_p);
+    } else if (!upipe_row_split_handle(upipe, uref, upump_p)) {
+        upipe_row_split_hold_input(upipe, uref);
+        upipe_row_split_block_input(upipe, upump_p);
+        /* Increment upipe refcount to avoid disappearing before all packets
+         * have been sent. */
+        upipe_use(upipe);
+    }
+}
+
+static bool upipe_row_split_handle(struct upipe *upipe, struct uref *uref,
+                                  struct upump **upump_p)
+{
+    struct upipe_row_split *upipe_row_split = upipe_row_split_from_upipe(upipe);
+    if (!upipe_row_split->ubuf_mgr)
+        return false;
+
+    size_t hsize, vsize;
+    uint8_t macropixel;
+    if (unlikely(!ubase_check(uref_pic_size(uref, &hsize, &vsize,
+                        &macropixel)))) {
+        upipe_warn(upipe, "dropping picture");
+        upipe_throw_error(upipe, UBASE_ERR_INVALID);
+        uref_free(uref);
+        return true;
+    }
+
+    uint64_t vsize_slice = upipe_row_split->slice_height;
+    if (vsize_slice > vsize)
+        vsize_slice = vsize;
+
+    for (uint64_t done = 0; done < vsize; done += vsize_slice) {
```

Comment:
I don't understand why you don't do simply:
new_ubuf = ubuf_dup(ubuf);
ubuf_pic_resize(new_ubuf, 0, done, -1, vsize_slice);
done += vsize_slice;